packages/lit-html/src/directives/spread.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {noChange, nothing, _Σ as litHtmlPrivate} from '../lit-html.js';
+import {
+  directive,
+  Directive,
+  DirectiveParameters,
+  PartInfo,
+  PartType,
+  Part,
+  ElementPart,
+} from '../directive.js';
+
+const {
+  _AttributePart: AttributePartImpl,
+  _PropertyPart: PropertyPartImpl,
+  _BooleanAttributePart: BooleanAttributePartImpl,
+  _EventPart: EventPartImpl,
+  // _ElementPart: ElementPartImpl,
+} = litHtmlPrivate;
+
+export interface SpreadValues {
+  readonly [name: string]: unknown;
+}
+
+class SpreadDirective extends Directive {
+  _previousValues: Set<string> = new Set();
+  _parts = new Map<string, Part>();
+
+  constructor(partInfo: PartInfo) {
+    super(partInfo);
+    if (partInfo.type !== PartType.ELEMENT) {
+      throw new Error(
+        'The `spread` directive must be used al element expression'
+      );
+    }
+  }
+
+  render(_values: SpreadValues) {
+    // TODO for SSR
```

Comment:
Do you have any nascent ideas here? Seems like we could go two ways here:
* Return an array of descriptors of the part types + values, and figure out how to reuse the `attribute-part` opcode handling to handle the per-part-specific stuff: https://github.com/lit/lit/blob/c8fe1d4c4a8b1c9acdd5331129ae3641c51d9904/packages/labs/ssr/src/lib/render-lit-html.ts#L612
* Return the array of _acutal_ part instances (iow, use the update code) that the `attribute-part` opcode handling can directly use, and figure out how to deal with the fact that you don't have an actual part to get the element to pass in `render` (which isn't technically a problem in SSR because _it_ doesn't need the element).


We do need to reuse code in the `attribute-part` opcode handling to ensure directives passed to spread work correctly.

Reply:
I'm not sure yet. I think something like just forwarding the object in an AttributeResult type might work, even if it's a bit weird to explain how that works or doesn't client-side.

I think figuring out hydration is the hard part here...